main/src/library/BPlusTree.flix

```diff
@@ -0,0 +1,763 @@
+                let rightSizeWithoutNew = rightSize - 1;
+                if(insertionPoint + 1 == mid) {
+                    // If insertionPoint is the first element copy all old keys and old elements and insert new child at 0.
+                    Array.copyInto(srcPos = mid - 1, dstPos = 0, len = rightSizeWithoutNew, src = leftNode->keys, rightKeys);
+                    Array.copyInto(srcPos = mid, dstPos = 1, len = rightSizeWithoutNew, src = leftNode->children, rightChildren);
+                    Array.put(insertedChild, 0, rightChildren);
+                    newKey
+                } else {
+                    let childIndexInRight = insertionPoint + 1 - mid;
+                    Array.copyInto(srcPos = mid, dstPos = 0, len = childIndexInRight - 1, src = leftNode->keys, rightKeys);
+                    Array.copyInto(srcPos = mid, dstPos = 0, len = childIndexInRight, src = leftNode->children, rightChildren);
+                    Array.put(insertedChild, childIndexInRight, rightChildren);
+                    Array.put(newKey, childIndexInRight - 1, rightKeys);
+                    // If insertionPoint is the last index we are done, otherwise insert rest of list.
+                    if(childIndexInRight != rightSizeWithoutNew) {
+                        Array.copyInto(srcPos = mid + childIndexInRight - 1, dstPos = childIndexInRight, len = rightSizeWithoutNew - childIndexInRight, src = leftNode->keys, rightKeys);
+                        Array.copyInto(srcPos = mid + childIndexInRight, dstPos = childIndexInRight + 1, len = rightSizeWithoutNew - childIndexInRight, src = leftNode->children, rightChildren)
+                    } else ();
+                    Array.get(mid - 1, leftNode->keys)
+                }
+            };
+            let rightNode = new Node @ rc {
+                keys = rightKeys,
+                children = rightChildren,
+                values = Array.empty(rc, 0),
+                lock = Lock.mkLock(rc),
+                size = rightSize,
+                parent = leftNode->parent,
+                next = leftNode->next,
+                isLeaf = false
+            };
+            leftNode->next = Some(rightNode);
+            Array.forEachWithIndex(i -> child ->
+                if (i < rightSize)
+                    child->parent = Some(rightNode)
+                else (), rightChildren);
+            Lock.lock(rightNode->lock); // pushKeyToParent expects a lock on the parent
+            lockParent(leftNode);
+            rightNode->parent = leftNode->parent;
+            pushKeyToParent(leftNode->parent, leftNode, rightNode, keyOfRight, order, rc)
+
+        ///
+        /// Push the key `newKey` to `parent`.
+        /// If `parent` is `Some(node)`, insert `newKey` into node if there is room. Otherwise split
+        /// `node` and return a new root node.
+        ///
+        /// If `parent` is `None`, create a new root node, insert `newKey` into it and return it.
+        ///
```

Comment:
order/size assumptions?

Reply:
I was just thinking that some inputs would crash the function, but i think its just that order should be >= 3?